### PR TITLE
Pass by reference instead of moves

### PR DESCRIPTION
**Pull request by conorpp**
Some stack space optimization.  Saves around 6-8KB in my tests.

### CODE REVIEW

File: src/macros.rs [hunk 1]
```diff
@@ -55,12 +55,17 @@ macro_rules! interchange {
                 static mut INTERCHANGES: [Option<$Name>; $N] = [NONE; $N];
                 static mut STATES: [u8; $N] = [0u8; $N];
                 unsafe {
-                    let mut cell: MaybeUninit<UnsafeCell<&'static mut Option<$Name>>> = MaybeUninit::uninit();
+                    // It's unnecessary to wrap interchange in Option<T> because of the state handling we are
+                    // already doing.  Since Option<T> must take atleast as much space as T, 
+                    // we transmute &mut Option<T> to &mut T.  There is another way to use MaybeUninit but it's not
+                    // yet stable.
+                    let underlying_data: &'static mut $Name = unsafe { core::mem::transmute::<_, _>(&mut INTERCHANGES[i]) };
```
Comment (nickray) [inline src/macros.rs]: I don't think this transmute is valid. At least I don't see an argument for it to be so; you seem to be making assumptions on the basis of "it must fit"?

Also, what is the "another way" with MaybeUninit?

Comment (nickray) [reply src/macros.rs]: I think this works and is less iffy:
```
#[derive(Clone, Debug, PartialEq)]    
pub enum $Name {    
    Request($REQUEST),    
    Response($RESPONSE),    
    None,    
}
(...)
static mut INTERCHANGES: [$Name; $N] = [$Name::None; $N];
(...)
cell.as_mut_ptr().write(UnsafeCell::new(&mut INTERCHANGES[i])); 
```
Can you check if it has the same stack-friendliness you desire? The actual `$Name` type isn't something the library user should interact with directly, and if we're already storing a discriminant to distinguish the Request/Response cases, having `$Name::None` around shouldn't be more than, like, one bit.

**Comment by nickray** [comment]
@conorpp I made some simplifications, hope they help :)

Good to remove the moves, back then I seem to have had some magical thinking about how "move" works; I honestly thought at that time that the compiler would "optimize everything out". Little did I know!

**Comment by conorpp** [comment]
I thought the same!  Sometimes it does, but I think in a lot of cases when passing/returning from functions it does not.

**Comment by nickray** [comment]
I think the "intermediate const initializer" trick should fix the Copy problem.

**Pull request merged**